src/software/ai/hl/stp/tactic/defense_shadow_enemy_tactic.cpp

```diff
@@ -0,0 +1,122 @@
+#include "software/ai/hl/stp/tactic/defense_shadow_enemy_tactic.h"
+
+#include "software/ai/hl/stp/action/move_action.h"
+#include "software/ai/hl/stp/action/stop_action.h"
+#include "software/ai/hl/stp/evaluation/calc_best_shot.h"
+#include "software/ai/hl/stp/evaluation/robot.h"
+#include "software/ai/hl/stp/tactic/tactic_visitor.h"
+#include "software/util/parameter/dynamic_parameters.h"
+
+DefenseShadowEnemyTactic::DefenseShadowEnemyTactic(
+    const Field &field, const Team &friendly_team, const Team &enemy_team,
+    const Ball &ball, bool ignore_goalie, double shadow_distance,
+    bool enemy_team_can_pass, bool loop_forever)
+    : Tactic(loop_forever),
+      field(field),
+      friendly_team(friendly_team),
+      enemy_team(enemy_team),
+      ball(ball),
+      ignore_goalie(ignore_goalie),
+      shadow_distance(shadow_distance),
+      enemy_team_can_pass(enemy_team_can_pass)
+{
+}
+
+std::string DefenseShadowEnemyTactic::getName() const
+{
+    return "Defense Shadow Enemy Tactic";
+}
+
+void DefenseShadowEnemyTactic::updateWorldParams(const Field &field,
+                                                 const Team &friendly_team,
+                                                 const Team &enemy_team, const Ball &ball)
+{
+    this->field         = field;
+    this->friendly_team = friendly_team;
+    this->enemy_team    = enemy_team;
+    this->ball          = ball;
+}
+
+void DefenseShadowEnemyTactic::updateControlParams(
+    const Evaluation::EnemyThreat &enemy_threat)
+{
+    this->enemy_threat = enemy_threat;
+}
+
+double DefenseShadowEnemyTactic::calculateRobotCost(const Robot &robot,
+                                                    const World &world)
+{
+    if (!enemy_threat)
+    {
+        return 0;
+    }
+    // Prefer robots closer to the enemy being shadowed
+    // We normalize with the total field length so that robots that are within the field
+    // have a cost less than 1
+    double cost = (robot.position() - enemy_threat->robot.position()).len() /
+                  world.field().totalXLength();
+    return std::clamp<double>(cost, 0, 1);
+}
+
+void DefenseShadowEnemyTactic::calculateNextIntent(IntentCoroutine::push_type &yield)
+{
+    MoveAction move_action = MoveAction();
+    StopAction stop_action =
+        StopAction(StopAction::ROBOT_STOPPED_SPEED_THRESHOLD_DEFAULT, true);
+    do
+    {
+        if (!enemy_threat)
```

Comment:
I understand that this case only occurs if there are no enemy robots, but is doing 'nothing' the best case here?

Reply:
In this case I think it is. Ideally we never get to this case, but without a "threat" there is no default intelligent place to go because there's just no information. I added a warning log to let us know when we are in this bad state.

Reply:
But during out mini STP-redesign (mostly coroutines) we should consider how we can design things to avoid these situations as much as possible. Obviously the best case is we never have to make this check because we can guarantee we have that data.

Reply:
I updated #773 to track this